### ISSUE TITLE
[UX] Redesign the Layouts page to make it easier to understand

### ISSUE THREAD

**Issue by philsward** [backdrop/backdrop-issues#4871]
Per discussions through several weekly meetings, this is the official issue to change the UI and UX of the Layouts page to be easier to understand when viewing it from a glance.

My proposal is to create more of a "card style" approach.

1) Separate each group type with a header
* Static Layout Groups
* Dynamic Layout Groups
* Default Layout Groups

2) Next, separate each layout group with its type as a "card" and allow multiple layouts with a group to live within its respective card.

3) Add ways to search and filter
* Add a search form (recommended to make it search via ajax)
* Add filters for each of the the layout group types

Rough draft mockup example:

![backdrop-layouts-mockup-2021-01-12-v1](https://user-images.githubusercontent.com/3436110/104409716-85676580-552c-11eb-8d48-d0042a58dbae.png)

[Related]
https://github.com/backdrop/backdrop-issues/issues/4868
https://github.com/backdrop/backdrop-issues/issues/4869

**Comment by stpaultim** [comment]
Well, I mentioned to @philsward that we have a few different issues about making the layouts page easier to understand, which is fine. If no one else does, I may need to create a META issue for this topic. 

I REALLY appreciate having a mock-up of the "card style" approach. I think the filters may be the more important change on this mock-up. The "card style" approach is probably an improvement over what we have now, but I'm not sure yet if its the best idea we can come up with. 

I look forward to hearing what others have in mind?

I'd certainly be interested in advocating for one of the layout page issues during the next release cycle, if there are any left for me to advocate for. ;-)

**Comment by philsward** [comment]
I think the one thing I would add that isn't already there, is an "Admin Layout Group". I'm torn on that idea though which is why I didn't add it plus it would probably be more difficult to implement.

**Comment by klonos** [comment]
> If no one else does, I may need to create a META issue for this topic. ...if there are any left for me to advocate for. ;-)

We seem to have plenty of them, so we may need someone to advocate for the meta-issue 😅 

I like the mockup with the grouping/separation of types of paths etc., but I see that we'll have trouble aligning the columns (so that they all have the same width).

**Comment by philsward** [comment]
> I see that we'll have trouble aligning the columns (so that they all have the same width).

As opposed to tables for the layout row data, would it make more sense to build them with flex or grid?

**Comment by klonos** [comment]
> ...would it make more sense to build them with flex or grid?

Or we might be able to get away with it by adding "pseudo-blank" separator table rows (no borders/background etc.). That should help keep all table columns with the same width 🤷

**Comment by philsward** [comment]
They way I pulled off the cards in my mock-up was to put each layout group in a div and the "header" (path) was also in a div, then each layout was left as a table.

**Comment by philsward** [comment]
Per @jenlampton in weekly meeting:

Suggestion to create two tabs, one for "site" the the other for "admin"

The idea is that we move the admin layouts into its own area away from the site layouts to reduce the clutter between front-end (site) and back-end (admin) display

Admin layouts rarely get added or changed so removing them from the site layout will streamline the main site layout focus.

**Comment by stpaultim** [comment]
Just a note that there was clear opposition to the use of tabs in the issue about single URL layouts versus dynamic URL layouts. Those concerns may or may not be relevant here.

https://github.com/backdrop/backdrop-issues/issues/4841#issuecomment-753017644

**Comment by jenlampton** [comment]
On the layouts list the Dashboard layout is almost always the one on top. And I'm finding that's the layout I want to customize the least, so having it at the top of the page is frustrating. 

Before I saw this issue, I wanted to split the layout listing into two sections. One, at the bottom of the page, that showed only the admin layouts. (For new sites, this would be Dashboard and Default Admin.) And the section at the top would show all the front-end layouts. This would put the layouts I want to configure most near the top of the page, and the ones that I ignore the most at the bottom. 

But this issue splits the layouts by group -- which I also like. I was unable to reconcile how to balance the group split with the admin/non-admin split. That's when @philsward mentioned tabs. In this case, I think it might be less offensive if we bury the admin layouts on the secondary tab, since those are the ones people probably aren't looking for anyway. 

We could also keep my original idea with the two tables (non-admin/admin) and still split both tables up in the same way.

**Comment by philsward** [comment]
I'm OK with two tabs to split site from admin.

I feel like this request is way less obtrusive than my original thoughts to split static vs dynamic and I agree with @jenlampton that most folks probably won't have an issue with this since admin layouts are rarely touched.

(Makes me wonder if this is a good idea for views too...)

Some thoughts: do we want to hard code the layout to be an admin vs non-admin group or should the group dynamically choose the tab based on the URL (if /admin, automatically put in admin tab, if non-admin automatically, put into site tab). I can see where this might cause confusion but it also means a layout can be converted from one to the other without cloning or complete rewrites. But what are the chances of that anyway...

If we go the tab route, what do we call them? Site and Admin are the obvious choices but are they explanatory enough? "Site Layouts" and "Admin Layouts" might make more sense. A mock-up with both might make this more clear.

**Comment by jenlampton** [comment]
> If we go the tab route, what do we call them? Site and Admin are the obvious choices 

I don't think we should use "Site" for anything, as that word can mean different things to different people. You seem to think it means only the front-end of the website, but I also see the admin layouts as belonging to my website, and for me "site" is an abbreviation of "website".

**Comment by docwilmot** [comment]
I dont think burying the default layout is necessarily beneficial. For a person with a simple site, the vast majority of pages would use that layout and most work would be done in there.

**Comment by stpaultim** [comment]
> I dont think burying the default layout is necessarily beneficial. For a person with a simple site, the vast majority of pages would use that layout and most work would be done in there.

@docwilmot - I agree. But, I don't think anyone has advocated burying the Default Layout. I think that @jenlampton and @philsward have talked about moving the admin layouts, including the `Default Administrative Layout`  to less prominent positions, because they are relatively rarely used. 

But, I don't think we're talking about burying the `Default Layout` - unless, I'm missing something.

**Comment by klonos** [comment]
Just wanted to introject this thought: the layout listing seems to be needing a way to separate layouts based on:

- admin vs. non-admin
- single-page (specific path) vs. "dynamic" (paths with `%` placeholders)

...while at the same time also group them by:

- path

This is too much grouping to do, and I cannot think of a single effective way to do it w/o complicating the UI, and making it look too "busy". I think that in such cases it may be best to list things alphabetically + add search/filtering/grouping controls at the top + make a specific thing "bubble up" the top (or drop to the bottom), based on a specific criterion. This is what we do in the site status report, where we list entries:

- first by severity (errors at the top -> then warnings -> then info/status rows at the bottom)
- entries within each group of severity are listed alphabetically

**Comment by docwilmot** [comment]
@stpaultim https://github.com/backdrop/backdrop-issues/issues/4871#issuecomment-760589433
> In this case, I think it might be less offensive if we bury the admin layouts on the secondary tab, since those are the ones people probably aren't looking for anyway.

**Comment by stpaultim** [comment]
@docwilmot - I'm confused about your last comment and what that quote is supposed to mean. Yes, your quote reflects exactly what I said. "Bury the admin layouts" not the Default layout. I don't think the default layout is an admin layout, is it?

I read "admin layouts" as layouts that effect administrative pages. The Default Layout effects content pages that don't have any other layout defined.

**Comment by docwilmot** [comment]
Yes, youre right. I was reading "admin" as "default". Time for coffee. Ignore.

**Comment by philsward** [comment]
> This is too much grouping to do, and I cannot think of a single effective way to do it w/o complicating the UI, and making it look too "busy". 

Well, the page needs to breathe. The table approach isn't allowing for that.

> I think that in such cases it may be best to list things alphabetically + add search/filtering/grouping controls at the top + make a specific thing "bubble up" the top (or drop to the bottom), based on a specific criterion. This is what we do in the site status report, where we list entries.

Layouts work a bit differently than info and status pages.

According to @jenlampton the layout is supposed to follow a rule order from bottom to top. Default is at the bottom because it matches everything (*) next you have dynamic (%) and if it matches, will override the default layout. Next you have the static layouts which override dynamic layouts.

Technically, /node/5 should be ABOVE /node/% but there's a bug where static paths are sorted alphabetically with dynamic and placed below them, hence the need to distinguish static from dynamic from default and ensure they follow the proper rules when being checked. And of course, @jenlampton mentioned this should never happen anyway because /node/5 should never even be allowed as a path group because you're supposed to use the dynamic with a visibility condition... So there's that. My vote is to allow both methods.

Dunno, definitely needs something 😉

**Comment by philsward** [comment]
> This is too much grouping to do, and I cannot think of a single effective way to do it w/o complicating the UI, and making it look too "busy".

@klonos I need to play with layouts more. I'm finding they don't work they way I expect and I need to pinpoint why they feel so foreign for me. They "should" be easy to understand, but the workflow of using them is all out of whack to me.

For example, if I have an /about node alias, I can't just override that node layout without first making the layout take complete control of that alias. Then I have to add the node as a visibility rule? Shouldn't the layout path BE the visibility condition? No, because layouts can be pages in their own right... The more I learn about layouts, the more confused I become... 😵‍💫

Regardless, the layout page needs to breathe somehow. It's "too busy" as it is 😉

**Comment by philsward** [comment]
@jenlampton I think I now understand why "/node/5" shouldn't work as a layout.

I'm beginning think "static layout" is a bad term. It's true, but is the wrong description. That or the fundamentals of layouts will need to change for it to be the correct description (guessing lots of opposition on that note)

There are:
* default paths
* dynamic paths

Then you have:
* layout pages

Two of them override based on the path, the other becomes an actual page with its own path.

(Thinking aloud)
With a layout page, nothing is being matched. Other than being a "layout", it has nothing else in common with dynamic and default layouts, therefore I don't feel like it should even live in the same space as dynamic and default layouts. 

I'll have to think on this a bit because "how I would fix the UX" would not be widely accepted based on the comments I've seen so far.

**Comment by philsward** [comment]
@jenlampton 
> I don't think we should use "Site" for anything, as that word can mean different things to different people.

Referring to #3122 we use the word "site" in the context of people viewing the front end of the website (non-admin areas). I agree that site can mean a lot of things, but in context to where the Layouts are ultimately displayed, and given we are specifying site vs admin, I think it is pretty clear that we mean "everything that isn't an admin page" when using the word site.

I guess I'm questioning whether we need to look at other areas that use it too?
